### ISSUE TITLE
关于CBN的实现

### ISSUE THREAD

**Issue by yyyyojiro** [automan000/Camera-based-Person-ReID#12]
您好，非常感谢并敬佩您的工作！我看到您项目trainer.py里传参数进self._forward()里的时候传的是一个列表，然后列表的不同元素是当前batch的不同相机的图片集，我想问问这样进行前传的时候会遍历列表单独算每个元素（相机）里图片的均值和标准差然后进行相机的正则化吗？谢谢

**Comment by automan000** [comment]
感谢关注。
首先，BN在训练阶段一直在使用batch mean和batch var对输入数据进行标准化，如果采样策略足够合理并且数据充足，batch mean和batch var是可以近似每一个Camera内的所有数据的统计量的。详情请见详情请见https://github.com/automan000/Camera-based-Person-ReID/issues/8。
其次，请关注我们自定义的[DataParallel](https://github.com/automan000/Camera-based-Person-ReID/blob/master/frameworks/training/data_parallel.py)，它可以保证训练阶段的每一次forward时，使用的样本都来自同一个Camera，也就是分别处理self._forward()的List中的每一个元素。

**Comment by yyyyojiro** [comment]
老师下午好，非常谢谢您的回复！我理解那部分的代码了！但是我还有三个小个问题想问您，第一个就是我看到您在测试的时候对每个相机采样3200张图片，那几乎整个query和gallery都在计算相机的统计信息，请问您有没有试过降低采样图片的数量，或者说是要达到论文中的效果的最小采样数是多少；第二个就是您在测试的时候"先self.model.train()再self.model.eval()"这个语句是不是为了计算BN的均值和方差啊，这样的操作之后只会对BN的统计信息产生影响吗？第三个就是您有没有试过在相机正则化的基础上将BN替换成其他的正则化方法比如Instance Normalization或者Group Normalization之类的。






------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "automan000/Camera-based-Person-ReID"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年4月29日(星期四) 下午2:34
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [automan000/Camera-based-Person-ReID] 关于CBN的实现 (#12)





 
感谢关注。
 首先，BN在训练阶段一直在使用batch mean和batch var对输入数据进行标准化，如果采样策略足够合理并且数据充足，batch mean和batch var是可以近似每一个Camera内的所有数据的统计量的。详情请见详情请见https://github.com/automan000/Camera-based-Person-ReID/issues/8。
 其次，请关注我们自定义的DataParallel，它可以保证训练阶段的每一次forward时，使用的样本都来自同一个Camera，也就是分别处理self._forward()的List中的每一个元素。

**Comment by automan000** [comment]
关于第一个问题，请见原文的Table.7，相关讨论见4.3 Ablation Study。对于采样数量，我们后续的研究发现数量对性能的影响并不是很显著，反而是采集到的图像质量（指能否正确地反映Camera对应的数据分布）对性能的影响更加显著。

对于第二点，是的。self.model.train()使得BN层开始更新moving average。同时，只要不进行回传操作，网络里的权重不会被更新。详见Pytorch里BN层的Python wrapper。

第三点，Instance Normalization可能无法与我们的方法直接结合。我们测试过IBN，详见原文Table.6 。Group Norm没有测试过。

**Comment by yyyyojiro** [comment]
好的，非常谢谢您的回复！我再去仔细阅读论文，祝您生活愉快，万事胜意！




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "automan000/Camera-based-Person-ReID"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年5月12日(星期三) 下午4:12
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [automan000/Camera-based-Person-ReID] 关于CBN的实现 (#12)





 
关于第一个问题，请见原文的Table.7，相关讨论见4.3 Ablation Study。对于采样数量，我们后续的研究发现数量对性能的影响并不是很显著，反而是采集到的图像质量（指能否正确地反映Camera对应的数据分布）对性能的影响更加显著。
 
对于第二点，是的。self.model.train()使得BN层开始更新moving average。同时，只要不进行回传操作，网络里的权重不会被更新。详见Pytorch里BN层的Python wrapper。
 
第三点，Instance Normalization可能无法与我们的方法直接结合。我们测试过IBN，详见原文Table.6 。Group Norm没有测试过。
